liesel/band.py

```diff
@@ -0,0 +1,264 @@
+    #         ),
+    #         jnp.zeros((bw, n)),
+    #     )
+
+    def up():
+        Ab = jnp.zeros((bw, n))
+        for k in range(bw):
+            Ab = Ab.at[k, k:].set(jnp.diag(full, k))
+        return jnp.flipud(Ab)
+
+    return jax.lax.cond(lower, lo, up)
+
+
+@partial(jax.jit, static_argnames="p")
+def to_band_2(full, p, lower=True):
+    """
+    Converts `full` to a `(p + 1) x n` band with bandwidth `p`.
+
+    ## Parameters
+
+    - `full`: A `n x n` matrix.
+    - `p`: The bandwidth `p`.
+    - `lower`: Whether to return the band in lower-diagonal ordered form.
+    """
+
+    n = full.shape[0]
+
+    padded = jnp.hstack([full, jnp.zeros([n, p])])
+
+    def f(i, row):
+        row = jax.lax.dynamic_slice(row, (i,), (bandwidth + 1,))
+        return i + 1, row
+
+    _, band = jax.lax.scan(f, 0, padded)
+
+    band = band.T
+    band = jax.lax.cond(lower, lambda x: x, to_upper, band)
+    return band
+
+
+# no @jax.jit
+def permute(full, lower=True):
+    """
+    Permutes `full` with the reverse Cuthill-McKee algorithm and returns the result
+    as a tuple of ((p x n) band, Permuation matrix P).
+
+    ## Parameters
+
+    - `full`: A full, symmetric(!) (n x n) matrix.
+    - `lower`: Whether to return the (p x n) band in lower-diagonal ordered form.
+    """
+    p = sps.csgraph.reverse_cuthill_mckee(sps.csr_matrix(full), symmetric_mode=True)
+    P = np.vstack([np.eye(1, full.shape[0], k) for k in p])
+    fullP = P @ full @ P.T
+
+    return to_band(fullP, bandwidth(fullP), lower), P
+
+
+@jax.jit
+def to_symm_full(band, lower=True):
+    """
+    Returns the symmetric full (n x n) matrix represented by `band`.
+
+    ## Parameters
+
+    - `band`: A (p x n) band with bandwidth p+1.
+    - `lower`: Whether `band` is in lower-diagonal ordered form.
+    """
+    a = to_utri_full(band, lower)
+    return a + jnp.tril(a.T, k=-1)
+
+
+@jax.jit
+def to_ltri_full(band, lower=True):
+    """
+    Returns the lower triangular full (n x n) matrix represented by `band`.
+
+    ## Parameters
+
+    - `band`: A (p x n) band with bandwidth p+1.
+    - `lower`: Whether `band` is in lower-diagonal ordered form.
+    """
+    return to_utri_full(band, lower).T
+
+
+@jax.jit
+def to_utri_full(band, lower=True):
+    """
+    Returns the upper triangular full (n x n) matrix represented by `band`.
+
+    ## Parameters
+
+    - `band`: A (p x n) band with bandwidth p+1.
+    - `lower`: Whether `band` is in lower-diagonal ordered form.
+    """
+    p, n = band.shape
+    p -= 1
+    return jax.lax.cond(
+        lower,
+        lambda: sum(jnp.diag(v[: n - k], k) for k, v in enumerate(band)),  # lower
+        lambda: sum(jnp.diag(band[k][p - k :], p - k) for k in range(p + 1)),  # upper
+    )
+
+
+@jax.jit
+def to_lower(band):
+    """
+    ## Parameters
+
+    - `band`: A (p x n) band in upper-diagonal ordered form.
+    """
+    return jnp.flipud(
+        jnp.array([jnp.roll(row, i - band.shape[0] + 1) for i, row in enumerate(band)])
```

Comment:
Could a similar approach as in `to_band_2` work here to get rid of the for loop?